Support for S3 bucket tracking.

johnworth opened the pull request:
##  Summary
  Added tracking for S3 bucket usage in analyses.

## Changes
* Added migration 000045 migration that adds tables for buckets, bucket_permissions, and job_buckets.
* The buckets table tracks the buckets created by the Discovery Environment. It does not contain auth information.
* The bucket_permissions table tracks a users permissions on a bucket. This is intended to allow sharing buckets along with analyses.
* The job_buckets table associates one or more buckets with each job. A bucket can be used in more than one job.
* The job_buckets table also tracks the mountpoint for the bucket in the analysis.
* Added down migrations as well.

Commit:
Adding buckets related tables.

Commit:
Add job_buckets table to 000045 up and down migrations

slr71 commented on migrations/000045_buckets.up.sql:
```diff
@@ -0,0 +1,96 @@
+    PRIMARY KEY (id),
+    FOREIGN KEY (bucket_id) REFERENCES buckets(id) ON DELETE CASCADE,
+    FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE,
+    FOREIGN KEY (created_by) REFERENCES users(id),
```
I'm not entirely sure I understand the purpose of this field. Is the idea that an administrator would have to grant permission for the bucket to be created? Also, what would happen if the administrator account was deleted?

johnworth replied:
The created_by field is the user who added the permission record and the user_id is the user on the receiving end. It's so a user can share a bucket with another user, not just if an admin shares it. You're right to point out the issue with the granting user deleting their account, I'm working on a fix now.

ianmcorvidae replied:
Reading it myself it looks like it's for tracking who granted what permission to what user, maybe? If there's multiple users who have admin-level permissions, it might be good to know who gave a given user read or write or admin access.

slr71 reviewed (approved):
I had one question, but this looks reasonable to me overall.

psarando reviewed (approved):
I appreciated the comments on the tables and columns.
LGTM 👍 (though I am also curious about Sarah's question).

ianmcorvidae commented on migrations/000045_buckets.up.sql:
```diff
@@ -0,0 +1,96 @@
+    FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE
+);
+
+CREATE INDEX idx_buckets_name ON buckets(name);
```
Either this line is not needed (less preferred, I think), or it should be `CREATE UNIQUE INDEX` and drop the `UNIQUE` above. Creating the unique index with a known name should make amending or dropping them later easier, since just adding `UNIQUE` in the `CREATE TABLE` definition lets postgres decide the index name.

Additionally, for this and the indices below, I believe you can do `CREATE INDEX IF NOT EXISTS` which is probably worth the characters for some additional idempotence.

johnworth replied:
Okay, those changes should be in now.

ianmcorvidae commented on migrations/000045_buckets.up.sql:
```diff
@@ -0,0 +1,96 @@
+    name text NOT NULL UNIQUE,
+    user_id uuid NOT NULL,
+    endpoint_url text NOT NULL,
+    region text,
```
I would sort of expect the `NOT NULL` to be swapped on these - a null `endpoint_url` probably just means "use standard Amazon S3 endpoints" but a null region is sort of hard to parse. Most tools for S3 that I've seen still expect a region to be passed even for a custom/non-Amazon S3 endpoint but just say to put in "N/A" or something, and we could just use the empty string. I don't see that there'd be a difference between an empty string region and a null region, so may as well only allow one of the two. Maybe the same applies to endpoint_url and we treat empty string as Amazon though, so maybe that's the answer.

johnworth replied:
I'm thinking for debugging/my sanity's sake, I'll keep the endpoint_url not null so we don't accidentally start depending on implicit behavior in a S3 tool/driver. I'll make the region not null as well for similar reasons, thanks for pointing that out.

johnworth replied:
The change to make the region NOT NULL is in.

ianmcorvidae commented on migrations/000045_buckets.up.sql:
```diff
@@ -0,0 +1,96 @@
+
+CREATE INDEX idx_buckets_name ON buckets(name);
+CREATE INDEX idx_buckets_user_id ON buckets(user_id);
+CREATE INDEX idx_buckets_deleted ON buckets(deleted);
```
This is probably a concern for later, but if we end up with a lot of deleted buckets we may eventually want indexes with the "WHERE NOT deleted" built into the index condition. But not worth adding until we're there I think.

ianmcorvidae commented on migrations/000045_buckets.up.sql:
```diff
@@ -0,0 +1,96 @@
+    FOREIGN KEY (bucket_id) REFERENCES buckets(id) ON DELETE CASCADE,
+    FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE,
+    FOREIGN KEY (created_by) REFERENCES users(id),
+    CONSTRAINT unique_bucket_user_permission UNIQUE (bucket_id, user_id)
```
As mentioned in the other table, I think this would be better done as a dedicated `CREATE UNIQUE INDEX` line so we control the index name that's used.

johnworth replied:
Will do, thanks.

johnworth replied:
That should be changed now.

ianmcorvidae reviewed (changes requested):
Got a few database-maintenance-streamlining sorts of comments. Basically looks fine but I'd definitely prefer those unique constraints be done as indices so we control the names in the DB a bit more.

Commit:
For bucket_permissions, allow granting users to be deleted without removing permissions

Commit:
Various fixes based on feedback

Makes the idx_buckets_name index creation explicitly a unique index.
Adds IF NOT EXISTS to the index creation statements.
Added a dedicated CREATE UNIQUE INDEX for the (bucket_id, user_id) fields on the bucket_permissions table.
Removes the NOT NULL constraint on the bucket_permissions.created_by field and have it get set to NULL if the user is deleted.

ianmcorvidae reviewed (approved):
LGTM! Thanks for the updates
